File: implementations/my_coa_colouring.py
import numpy as np

adj_matrix = [  [0, 1, 1, 1, 1, 1],
				[1, 0, 1, 1, 1, 1],
				[1, 1, 0, 1, 1, 1],
				[1, 1, 1, 0, 1, 1],
				[1, 1, 1, 1, 0, 1],
				[1, 1, 1, 1, 1, 0]]
n = len(adj_matrix)

def f(x):
	"""Returns the number of colours used by a colouring"""
	return x.max() + 1

def d(x, y):
	"""Returns the distance between x and cuckoo y"""
	ret = 0
	for i in range(n):
		if x[i] != y[i]:
			ret += 1
	return ret

def d_bar_sum(S1, S2):
	num = 0
	for i in S1:
		for j in S2:
			num += dist_matrix[i][j]
	return num

def tri_dist(i, j, S, dbss, clusters):
	"""Finds the triangular distance between cuckoo i and cluster j"""
	if len(clusters[j]) == 0:
		return -1
	return 2*d_bar_sum([i], clusters[j])/len(clusters[j]) - dbss[j]/(len(clusters[j]) * len(clusters[j]))

k = 3
def goal_point():
	populate_dist_matrix()
	S = np.random.randint(0, k, (n_pop))
	clusters = [[] for i in range(k)]
	for i in range(n_pop):
		clusters[S[i]].append(i)
	db_sum_self = np.zeros((k), dtype=int)  # Stores the sum of distance from every member to every other member (for calculation of d_bar(i, i))
	for i in range(k):
		for j in range(len(clusters[i])):
			for l in range(j):
				db_sum_self[i] += 2*dist_matrix[clusters[i][j]][clusters[i][l]]
	converged = False
	while not converged:
		converged = True
		for i in range(n_pop):
			cluster = -1
			tdc = -1
			for j in range(k):
				td = tri_dist(i, j, S, db_sum_self, clusters)
				if (cluster == -1 or td < tdc) and td != -1:
					cluster = j
					tdc = td
			if cluster != S[i]:
				clusters[S[i]].remove(i)
				clusters[cluster].append(i)
				converged = False
				for j in clusters[S[i]]:
					db_sum_self[S[i]] -= 2*dist_matrix[j][i]
				for j in clusters[cluster]:
					db_sum_self[cluster] += 2*dist_matrix[i][j]
				S[i] = cluster
	cluster_fitness = [[f(cuckoos[c]) for c in clusters[i]] for i in range(k)]
	best_cluster = -1
	best_cluster_mean_fitness = 0
	for i in range(k):
		if len(cluster_fitness[i]) > 0 and (best_cluster == -1 or np.mean(cluster_fitness[i]) < best_cluster_mean_fitness):
			best_cluster = i
			best_cluster_mean_fitness = np.mean(cluster_fitness[i])
	gp = -1
	for i in range(n_pop):
		if S[i] == best_cluster and (gp == -1 or f(cuckoos[i]) < f(cuckoos[gp])):
			gp = i
	return gp

def valid(v, c, col):
	"""Returns whether or not vertex v can be coloured c in col"""
	for i in range(n):
		if adj_matrix[v][i] == 1 and col[i] == c:
			return False
	return True

def generate_cuckoo():
	colouring = np.array([-1 for i in range(n)])
	order = np.array([i for i in range(n)])
	np.random.shuffle(order)
	for v in order:
		c = 0
		while True:
			if valid(v, c, colouring):
				colouring[v] = c
				break
			c += 1
	return colouring

def get_egg(x, elr):
	"""Returns a valid colouring within distance elr from x"""
	num = np.random.randint(0, elr + 1)
	# colour num vertices greedily
	col = x.copy()
	for i in range(num):
		v = np.random.randint(0, n)
		for c in range(n):
			if c != x[v] and valid(v, c, col):
				col[v] = c
				break
	return col

def minimise(col):
	"""Reduces the max number of a colouring (while keeping all colours consistent)"""
	colours = list(np.unique(col, axis=0))
	for i in range(n):
		col[i] = colours.index(col[i])

def migrate(x, y):
	"""Migrates x towards y (in place)"""
	# Generate random proportion of distance to travel
	r = np.random.uniform(0, 1)
	# Let I be the vertices on which x and y disagree
	I = np.array([i for i in range(n) if x[i] != y[i]])
	for i in range(int(r * len(I))):
		v = I[i]
		# Replace i with y's colouring for i
		x[v] = y[v]
		# Clean up col
		for j in range(i + 1, n):
			# For every conflicting edge (v,j), replace j's colour with the smallest legal colour different than y[j]
			if adj_matrix[j][v] == 1 and x[j] == x[v]:
				c = 0
				while True:
					if c != y[j] and valid(j, c, x):
						x[j] = c
						break
					c += 1
	#minimise(x)  # Not sure if necessary

def populate_dist_matrix():
	for i in range(n_pop):
		for j in range(i):
			dist_matrix[i][j] = d(cuckoos[i], cuckoos[j])
			dist_matrix[j][i] = dist_matrix[i][j]

alpha = 10
n_pop = 5
n_max = 50
num_iterations = 3000
p = 0.1

cuckoos = np.array([generate_cuckoo() for i in range(n_pop)])

dist_matrix = np.zeros((n_max, n_max), dtype=int)
populate_dist_matrix()

for t in range(num_iterations):
	num_eggs = np.random.randint(5, 21, (n_pop))
	tot_eggs = num_eggs.sum()
	eggs = np.zeros((tot_eggs, n), dtype=int)
	egg = 0
	for i in range(n_pop):
		elr = alpha * num_eggs[i]/tot_eggs * n  # I've just put n instead of (v_hi - v_lo)
		for j in range(num_eggs[i]):
			eggs[egg] = get_egg(cuckoos[i], elr)
			egg += 1
	eggs = eggs[np.array([f(e) for e in eggs]).argsort()]
	cuckoos = np.append(eggs[:int(tot_eggs * (1 - p))], cuckoos, axis=0)
	if len(cuckoos) > n_max:
		cuckoos = cuckoos[np.array([f(c) for c in cuckoos]).argsort()]
		cuckoos.resize((n_max, n))
	n_pop = len(cuckoos)
	gp = goal_point()
	for i in range(n_pop):
		migrate(cuckoos[i], cuckoos[gp])

cuckoos = cuckoos[np.array([f(c) for c in cuckoos]).argsort()]
print(cuckoos[0])

File: implementations/my_coa_colouring/my_coa_colouring/my_coa_colouring.cpp
#define _USE_MATH_DEFINES
#define _SECURE_SCL 0
#define NUM_VERTICES 100
#define N_MAX 50

#include <iostream>
#include <array>
#include <vector>
#include <math.h>
#include <chrono>
#include <algorithm>
#include <boost/random/uniform_real_distribution.hpp>
#include <boost/random/uniform_int_distribution.hpp>
#include <boost/random/mersenne_twister.hpp>

// Running times (eggs 5 to 20, pop 5 to 50, iterations 3000)
	// 10 vertices: 17.94 secs
	

using namespace std;
using namespace boost::random;

typedef struct Cuckoo {
	int cuckoo[NUM_VERTICES];
	int fitness;
};

vector<int> edge_list[NUM_VERTICES];/* = {
	{1, 4, 5},
	{0, 2, 6},
	{1, 3, 7},
	{2, 4, 8},
	{3, 1, 9},
	{0, 7, 8},
	{1, 8, 9},
	{2, 5, 8},
	{3, 5, 6},
	{4, 6, 7}
};*/

int n_pop = 5;

const float alpha = 1;
const int num_iterations = 3000;
const float p = 0.1;
const int min_eggs = 1;
const int max_eggs = 1;

Cuckoo cuckoos[N_MAX];

int dist_matrix[N_MAX][N_MAX];

int num_eggs[N_MAX];
Cuckoo eggs[N_MAX * max_eggs];

mt19937 seed;
uniform_real_distribution<float> uni(0, 1);
uniform_int_distribution<int> random_vertex(0, NUM_VERTICES - 1);
uniform_int_distribution<int> random_egg_num(min_eggs, max_eggs);

void make_graph(float edge_probability) {
	for (int i = 0; i < NUM_VERTICES; i++) {
		for (int j = 0; j < i; j++) {
			if (uni(seed) < edge_probability) {
				edge_list[i].push_back(j);
				edge_list[j].push_back(i);
			}
		}
	}
}

int f(int * x) {
	int max = 0;
	for (int i = 0; i < NUM_VERTICES; i++) {
		if (x[i] > max) {
			max = x[i];
		}
	}
	return max + 1;
}

int d(int x, int y) {
	int ret = 0;
	for (int i = 0; i < NUM_VERTICES; i++) {
		if (cuckoos[x].cuckoo[i] != cuckoos[y].cuckoo[i]) {
			ret++;
		}
	}
	return ret;
}

int d_bar_sum(vector<int> S1, vector<int> S2) {
	int ret = 0;
	for (auto it = S1.begin(); it != S1.end(); ++it) {
		for (auto jit = S2.begin(); jit != S2.end(); ++jit) {
			ret += dist_matrix[*it][*jit];
		}
	}
	return ret;
}

float tri_dist(int i, int j, int * dbss, vector<int> * clusters) {
	if (clusters[j].size() == 0) {
		return -1;
	}
	return 2 * d_bar_sum({ i }, clusters[j]) / clusters[j].size() - dbss[j] / (clusters[j].size() * clusters[j].size());
}

void populate_dist_matrix() {
	for (int i = 0; i < n_pop; i++) {
		for (int j = 0; j < i; j++) {
			dist_matrix[i][j] = d(i, j);
			dist_matrix[j][i] = dist_matrix[i][j];
		}
	}
}

const int k = 3;
int S[N_MAX];
vector<int> clusters[k];
int db_sum_self[k];
uniform_int_distribution<int> random_cluster(0, k - 1);

int goal_point() {
	populate_dist_matrix();
	// Assign each cuckoo to a random cluster
	for (int i = 0; i < k; i++) {
		clusters[i].clear();
	}
	for (int i = 0; i < n_pop; i++) {
		S[i] = random_cluster(seed);
		clusters[S[i]].push_back(i);
	}
	// Initialise db_sum_self
	for (int i = 0; i < k; i++) {
		db_sum_self[i] = 0;
		for (auto it = clusters[i].begin(); it != clusters[i].end(); ++it) {
			for (auto jit = clusters[i].begin(); jit < it; ++jit) {
				db_sum_self[i] += 2 * dist_matrix[*it][*jit];
			}
		}
	}
	bool converged = false;
	while (!converged) {
		converged = true;
		for (int i = 0; i < n_pop; i++) {
			int cluster = -1;
			float tdc = -1;
			for (int j = 0; j < k; j++) {
				float td = tri_dist(i, j, db_sum_self, clusters);
				if ((cluster == -1 || td < tdc) && td != -1) {
					cluster = j;
					tdc = td;
				}
			}
			if (cluster != S[i]) {
				// Remove i from clusters[S[i]]
				for (auto it = clusters[S[i]].begin(); it != clusters[S[i]].end(); ++it) {
					if (*it == i) {
						clusters[S[i]].erase(it);
						break;
					}
				}
				clusters[cluster].push_back(i);
				converged = false;
				for (auto it = clusters[S[i]].begin(); it != clusters[S[i]].end(); ++it){
					db_sum_self[S[i]] -= 2 * dist_matrix[i][*it];
				}
				for (auto it = clusters[cluster].begin(); it != clusters[cluster].end(); ++it) {
					db_sum_self[cluster] += 2 * dist_matrix[i][*it];
				}
				S[i] = cluster;
			}
		}
	}
	int best_cluster = -1;
	float best_cluster_mean_fitness = 0;
	for (int i = 0; i < k; i++) {
		float mean_cluster_fitness = 0;
		for (auto it = clusters[i].begin(); it != clusters[i].end(); ++it){
			mean_cluster_fitness += cuckoos[*it].fitness;
		}
		mean_cluster_fitness /= clusters[i].size();
		if (clusters[i].size() > 0 && (best_cluster == -1 || mean_cluster_fitness < best_cluster_mean_fitness)) {
			best_cluster = i;
			best_cluster_mean_fitness = mean_cluster_fitness;
		}
	}
	int gp = clusters[best_cluster][0];
	for (auto it = clusters[best_cluster].begin(); it != clusters[best_cluster].end(); ++it) {
		if (cuckoos[*it].fitness < cuckoos[gp].fitness) {
			gp = *it;
		}
	}
	return gp;
}

bool valid(int v, int c, int col) {
	for (auto it = edge_list[v].begin(); it != edge_list[v].end(); ++it){
		if (cuckoos[col].cuckoo[*it] == c) {
			return false;
		}
	}
	return true;
}

int order[NUM_VERTICES];

void generate_cuckoo(int cuckoo) {
	// Choose a random ordering of vertices
	random_shuffle(begin(order), end(order));
	for (int v : order) {
		int c = 0;
		while (true) {
			if (valid(v, c, cuckoo)) {
				cuckoos[cuckoo].cuckoo[v] = c;
				break;
			}
			c++;
		}
	}
}

void get_egg(int cuckoo, float elr) {
	int num = uniform_int_distribution<int>(0, elr)(seed);
	for (int i = 0; i < num; i++) {
		int v = random_vertex(seed);
		int c = 0;
		while (true) {
			if (c != cuckoos[cuckoo].cuckoo[v] && valid(v, c, cuckoo)) {
				cuckoos[cuckoo].cuckoo[v] = c;
				break;
			}
			c++;
		}
	}
}

bool compare_cuckoos(Cuckoo c1, Cuckoo c2) {
	return c1.fitness < c2.fitness;
}

bool reverse_compare_cuckoos(Cuckoo c1, Cuckoo c2) {
	return c1.fitness > c2.fitness;
}

int I[NUM_VERTICES];
void migrate(int x, int y) {
	float r = uni(seed);
	// Populate a vector I with all vertices on which x and y disagree
	int I_length = 0;
	for (int i = 0; i < NUM_VERTICES; i++) {
		if (cuckoos[x].cuckoo[i] != cuckoos[y].cuckoo[i]) {
			I[I_length] = i;
			I_length++;
		}
	}
	for (int i = 0; i < r * I_length; i++){
		int v = I[i];
		// Assign x y's colour for *it
		cuckoos[x].cuckoo[v] = cuckoos[y].cuckoo[v];
		// Clean up x
		for (int j = i + 1; j < NUM_VERTICES; j++) {
			// For every conflicting edge (v, *jit), replace *jit's colour with the smallest legal colour different than y[*jit]
			if (cuckoos[x].cuckoo[j] == cuckoos[x].cuckoo[v] && find(edge_list[v].begin(), edge_list[v].end(), j) != edge_list[v].end()) {
				int c = 0;
				while (true) {
					if (c != cuckoos[y].cuckoo[v] && valid(j, c, x)) {
						cuckoos[x].cuckoo[j] = c;
						break;
					}
					c++;
				}
			}
		}
	}
}

int main(){  // I'm pretty sure the bottleneck is clustering
	make_graph(0.5);
	// Populate order array for generating cuckoos
	for (int i = 0; i < NUM_VERTICES; i++) {
		order[i] = i;
	}
	
	// Populate cuckoo array
	for (int i = 0; i < n_pop; i++) {
		generate_cuckoo(i);
		cuckoos[i].fitness = f(cuckoos[i].cuckoo);
	}
	auto start = chrono::high_resolution_clock::now();
	for (int t = 0; t < num_iterations; t++) {
		// Lay eggs
		int tot_eggs = 0;
		int egg = 0;
		for (int i = 0; i < n_pop; i++) {
			num_eggs[i] = random_egg_num(seed);
			tot_eggs += num_eggs[i];
		}
		for (int i = 0; i < n_pop; i++) {
			float elr = alpha * (num_eggs[i] / (float)tot_eggs) * NUM_VERTICES;
			for (int j = 0; j < num_eggs[i]; j++) {
				copy(begin(cuckoos[i].cuckoo), end(cuckoos[i].cuckoo), begin(eggs[egg].cuckoo));
				get_egg(egg, elr);
				eggs[egg].fitness = f(eggs[egg].cuckoo);
				egg++;
			}
		}
		// Kill a fraction p of the worst eggs
		sort(begin(eggs), begin(eggs) + tot_eggs, compare_cuckoos);
		tot_eggs *= (1 - p);
		if (n_pop + tot_eggs > N_MAX) {  // See if we can get away without using copy here
			// Add eggs until n_pop = N_MAX
			for (int i = 0; i < N_MAX - n_pop; i++) {
				copy(begin(eggs[i].cuckoo), end(eggs[i].cuckoo), begin(cuckoos[n_pop + i].cuckoo));
				cuckoos[n_pop + i].fitness = eggs[i].fitness;
			}
			// Let best remaining eggs replace worst cuckoos
			sort(begin(cuckoos), begin(cuckoos) + n_pop, reverse_compare_cuckoos);
			for (int i = N_MAX - n_pop; i < tot_eggs; i++) {
				if (eggs[i].fitness < cuckoos[i - N_MAX + n_pop].fitness) {
					copy(begin(eggs[i].cuckoo), end(eggs[i].cuckoo), begin(cuckoos[i - N_MAX + n_pop].cuckoo));
					cuckoos[i - N_MAX + n_pop].fitness = eggs[i].fitness;
				}
			}
			n_pop = N_MAX;
		}else{
			for (int i = 0; i < tot_eggs; i++) {
				n_pop++;
				copy(begin(eggs[i].cuckoo), end(eggs[i].cuckoo), begin(cuckoos[n_pop].cuckoo));
				cuckoos[n_pop].fitness = eggs[i].fitness;
			}
		}
		int gp = goal_point();
		//int gp = 0;
		for (int i = 0; i < n_pop; i++) {
			migrate(i, gp);
			cuckoos[i].fitness = f(cuckoos[i].cuckoo);
		}
		//cout << chrono::duration_cast<chrono::microseconds>(chrono::high_resolution_clock::now() - start).count() << endl;
	}
	for (int i = 0; i < NUM_VERTICES; i++) {
		cout << cuckoos[0].cuckoo[i] << " ";
	}
	cout << endl << "Time taken: " << chrono::duration_cast<chrono::microseconds>(chrono::high_resolution_clock::now() - start).count();
	return 0;
}